File: Grind75/coinChange.py
def coinChange(coins: List[int], amount: int) -> int:
    """
    DP approach: find the fewest number of coins needed to make up the given amount.
    
    dp[i] will hold the minimum coins needed to make sum = i.
    """
    # Use amount+1 as an "infinite" sentinel value (no solution can use more than amount coins of 1)
    INF = amount + 1
    
    # Initialize dp array:
    # dp[0] = 0 (zero coins needed for sum 0)
    # dp[1..amount] = INF (unknown yet, start as "infinite")
    dp = [0] + [INF] * amount

    # Build up dp from 1 to target amount
    for curr_sum in range(1, amount + 1):
        # Try every coin to see if it can contribute to curr_sum
        for coin in coins:
            if coin <= curr_sum:
                # If we use this coin, we need 1 coin + best for (curr_sum - coin)
                dp[curr_sum] = min(dp[curr_sum], dp[curr_sum - coin] + 1)
                # dp[curr_sum - coin] is already the fewest coins for that sub-sum

    # If dp[amount] is still INF, no combination made that sum
    return dp[amount] if dp[amount] != INF else -1


def coinChangeBFS(coins: List[int], amount: int) -> int:
    """
    BFS approach: treat each "remaining amount" as a node in a graph.
    Each edge subtracts a coin, and BFS levels equal number of coins used.
    """
    # Edge case: zero amount needs zero coins
    if amount == 0:
        return 0

    # Keep track of which remainders we've visited to avoid repeats
    visited = {amount}
    queue = deque([amount])
    level = 0  # number of coins used so far

    # Standard BFS loop
    while queue:
        level += 1
        # Process all nodes (remainders) at this BFS level
        for _ in range(len(queue)):
            remaining = queue.popleft()
            # Try subtracting each coin denomination
            for coin in coins:
                next_amount = remaining - coin
                if next_amount == 0:
                    # Found a path to exactly 0: return coins used (current level)
                    return level
                # If it's still positive and not seen before, enqueue it
                if next_amount > 0 and next_amount not in visited:
                    visited.add(next_amount)
                    queue.append(next_amount)
    # Exhausted all possibilities without hitting 0: no solution
    return -1

File: Grind75/coinChange.js
function coinChange(coins, amount) {
  // dp[i] will be the fewest coins needed to make amount i
  const dp = new Array(amount + 1).fill(Infinity);
  dp[0] = 0; // zero coins to make amount 0

  // For each sub‑amount from 1 to amount...
  for (let i = 1; i <= amount; i++) {
    // try every coin
    for (const coin of coins) {
      if (coin <= i && dp[i - coin] !== Infinity) {
        dp[i] = Math.min(dp[i], dp[i - coin] + 1);
      }
    }
  }

  return dp[amount] === Infinity ? -1 : dp[amount];
}

// BFS solution
function coinChange(coins, amount) {
  if (amount === 0) return 0;

  // Sort descending to try larger coins first (optional optimization)
  coins.sort((a, b) => b - a);

  const visited = new Array(amount + 1).fill(false);
  const queue = [amount];
  visited[amount] = true;
  let steps = 0;

  while (queue.length) {
    steps++;
    const len = queue.length;
    for (let i = 0; i < len; i++) {
      const curr = queue.shift();
      for (const coin of coins) {
        const next = curr - coin;
        if (next === 0) {
          // found exact change in `steps` coins
          return steps;
        }
        // only enqueue positive, unvisited remainders
        if (next > 0 && !visited[next]) {
          visited[next] = true;
          queue.push(next);
        }
      }
    }
  }

  // never reached 0
  return -1;
}
